### PR TITLE
Add the ability to change secrets provider

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #481

https://asciinema.org/a/ftLCUrN0NPMCTFQo83T1VYrGn

### CODE REVIEW

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}),
+	}
+
+	cmd.PersistentFlags().StringVar(
+		&secretsProvider, "secrets-provider", "", possibleSecretsProviderChoices)
```
[the remark this answers is not in the document]
Comment (stack72) [reply pkg/cmd/pulumi/stack_change_secrets_provider.go:136]: Great idea - it's not an optional argument so we shouldn't treat it as such

**Comment by stack72** [comment]
@justinvp this is all yours again!

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+				return err
+			}
+
+			// Get the current backend
+			b, err := currentBackend(opts)
+			if err != nil {
+				return err
+			}
+
+			// Get the current stack and it's project
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:70]: Nit:

```suggestion
			// Get the current stack and its project
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			currentStack, err := requireStack("", false, opts, true /*setCurrent*/)
+			if err != nil {
+				return err
+			}
+			currentProjectStack, err := loadProjectStack(currentStack)
+			if err != nil {
+				return err
+			}
+
+			// Build encrypter and decrypter based on the existing secrets provider
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:80]: Nit:

```suggestion
			// Build decrypter based on the existing secrets provider
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+			return err
+		}
+	}
+
+	err = saveProjectStack(currentStack, reloadedProjectStack)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:191]:
```suggestion
	return saveProjectStack(currentStack, reloadedProjectStack)
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	reloadedProjectStack, err := loadProjectStack(currentStack)
+	if err != nil {
+		return err
+	}
+
+	for key, val := range newProjectConfig {
+		err = reloadedProjectStack.Config.Set(key, val, false)
+		if err != nil {
+			return err
+		}
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:183]:
```suggestion
		if err := reloadedProjectStack.Config.Set(key, val, false); err != nil {
			return err
		}
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		Version:    apitype.DeploymentSchemaVersionCurrent,
+		Deployment: bytes,
+	}
+
+	// Import the newly changes Deployment
+	if err = currentStack.ImportDeployment(ctx, &dep); err != nil {
+		return err
+	}
+
+	return err
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:156]:
```suggestion
	// Import the newly changes Deployment
	return currentStack.ImportDeployment(ctx, &dep)
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	// Get the newly created secrets manager for the stack
+	newSecretsManager, err := getStackSecretsManager(currentStack)
+	if err != nil {
+		return err
+	}
+
+	// Reserialize the Snapshopshot with the NewSecrets Manager
+	reserializedDeployment, err := stack.SerializeDeployment(snap, newSecretsManager, false)
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:136]: Nit:

```suggestion
	reserializedDeployment, err := stack.SerializeDeployment(snap, newSecretsManager, false /*showSecrets*/)
```

File: pkg/cmd/pulumi/stack_change_secrets_provider.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+				return err
+			}
+
+			// Fixup the checkpoint
+			err = migrateCheckpointToNewSecretsProvider(commandContext(), currentStack)
+			if err != nil {
+				return err
+			}
+
+			return nil
```
Comment (justinvp) [inline pkg/cmd/pulumi/stack_change_secrets_provider.go:111]:
```suggestion
			return migrateCheckpointToNewSecretsProvider(commandContext(), currentStack)
```

**Comment by stack72** [comment]
Thanks for the review here @justinvp - all changes made as suggested!

When we get a green build, I will merge it

**Pull request merged**